go/providers.rst

```diff
@@ -1,34 +1,172 @@
-Go toolchains
-=============
+Go providers
+============
+
+.. _providers: https://docs.bazel.build/versions/master/skylark/rules.html#providers
+
+.. _go_library: core.rst#go_library
+.. _go_binary: core.rst#go_binary
+.. _go_test: core.rst#go_test
+.. _cc_library: https://docs.bazel.build/versions/master/be/c-cpp.html#cc_library
+.. _flatbuffers: http://google.github.io/flatbuffers/
+.. _static linking: modes.rst#building-static-binaries
+.. _race detector: modes.rst#using-the-race-detector
+.. _runfiles: https://docs.bazel.build/versions/master/skylark/lib/runfiles.html
 
 .. role:: param(kbd)
 .. role:: type(emphasis)
 .. role:: value(code)
 .. |mandatory| replace:: **mandatory value**
 
+
+The providers_ are the outputs of the rules, you generaly get them by having a dependancy on a rule,
+and then asking for a provider of a specific type.
+
 .. contents:: :depth: 2
 
 -----
 
 Design
 ------
 
-**TODO**: More information
+The go providers are designed primarily for the efficiency of the go rules, the information they
+share is mostly there because it is required for the core rules to work.
+
+All the providers are designed to hold only immutable data. This is partly because its a cleaner
+design choice to be able to assume a provider will never change, but also because only immutable
+objects are allowed to be stored in a depset, and it's really useful to have depsets of providers.
+Specifically the :param:`direct` and :param:`transitive` fields on GoLibrary_ only work because
+it is immutable.
 
 API
 ---
 
 GoLibrary
 ~~~~~~~~~
 
-**TODO**: More information
+This is the provider exposed by the go_library_ rule, or anything that wants to behave like one.
+It provides all the information requried to use the library as a dependancy, for other libraries,
+binaries or tests.
+
++--------------------------------+-----------------------------------------------------------------+
+| **Name**                       | **Type**                                                        |
++--------------------------------+-----------------------------------------------------------------+
+| :param:`importpath`            | :type:`string`                                                  |
++--------------------------------+-----------------------------------------------------------------+
+| The import path for this library.                                                                |
++--------------------------------+-----------------------------------------------------------------+
+| :param:`direct`                | :type:`depset(GoLibrary)`                                       |
++--------------------------------+-----------------------------------------------------------------+
+| The direct depencancies of the library.                                                          |
```

Comment:
Typo here too. (Sorry for intruding on the review).

Reply:
No intrusion, thanks!
Turns out there were quite a few, I just can't spell that word or any of its derivatives